compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java

```diff
@@ -489,30 +485,82 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
      *   ;
      */
     private boolean parseCallSuffix() {
-        if (parseCallWithClosure()) {
-            // do nothing
-        }
-        else if (at(LPAR)) {
-            parseValueArgumentList();
-            parseCallWithClosure();
-        }
-        else if (at(LT)) {
+        if (at(LT)) {
             PsiBuilder.Marker typeArgumentList = mark();
-            if (myKotlinParsing.tryParseTypeArgumentList(TYPE_ARGUMENT_LIST_STOPPERS)) {
+
+            if (!myKotlinParsing.tryParseTypeArgumentList(/* strict */false)) {
+                // in a non-strict/mixed environment it is very finicky to decide whether an erroneous
+                // type argument list is a call-suffix or a comparison. Even close-to-correct cases
+                // such as "<identifier, ()>" can still be considered as either. The only trivial case
+                // is "<>", which is already handled by tryParseTypeArgumentList(boolean).
+                typeArgumentList.rollbackTo();
+
+                return false;
+            }
+
+            // The definite cases were the call-suffix takes precedence
+
+            // <...> (
+            if (isAtValueArgumentList()) {
                 typeArgumentList.done(TYPE_ARGUMENT_LIST);
-                if (!myBuilder.newlineBeforeCurrentToken() && at(LPAR)) parseValueArgumentList();
+
+                parseValueArgumentList();
                 parseCallWithClosure();
+                return true;
             }
-            else {
-                typeArgumentList.rollbackTo();
-                return false;
+
+            // <...> {
+            if (isAtAnnotatedLambda()) {
+                typeArgumentList.done(TYPE_ARGUMENT_LIST);
+
+                parseCallWithClosure();
+                return true;
             }
-        }
-        else {
+
+            // <...>::
+            if (at(COLONCOLON)) {
```

Comment:
I'm going to a lot of test cases for all sorts of corner cases. I just didn't add them yet, because I wanted to work out the details of the approach first, without having to adjust the ranges within the tests over and over.